### PR TITLE
Produce diagnostics for KnownValueTypes that have annotations in its members

### LINKED ISSUE #2339: [DAM analyzer] Port DAM binding logic to analyzer
The Roslyn analyzer needs to be able to produce warnings for reflection-access to reflection-annotated methods. Whenever we see a reflection pattern that accesses, for example, `PublicMethods` on a particular type, we need to be able to find all public methods on the type.

The binding logic from https://github.com/dotnet/linker/blob/main/src/linker/Linker.Dataflow/DynamicallyAccessedMembersBinder.cs should be ported to work with Roslyn, and made to produce these warnings. To do this, I would start by enabling a few tests from https://github.com/dotnet/linker/blob/main/test/Mono.Linker.Tests.Cases/DataFlow/AnnotatedMembersAccessedViaReflection.cs in the `feature/damAnalyzer` branch (which has early support for DynamicallyAccessedMembersAttribute).

### PR DESCRIPTION
**Pull request by tlakollo**
Add KnownTypeValue as a subclass of ValueWithDynamicallyAccessedMembers
Add IL2110 and IL2111 to shared Diagnostic ids
Add test in analyzer for now since intrinsics are still not handled
Delete the use of context in the binder since the context is never used
Produce diagnostics for KnownValueTypes that have annotations in its members

Fixes https://github.com/dotnet/linker/issues/2410
Fixes #2339

### CODE REVIEW

**Review by sbomer — COMMENTED** [review]
Looking good! Would it be reasonable to turn on the test `AnnotatedMembersAccessedViaReflection` and use `ProducedBy` to skip the ones that rely on intrinsics? It would be good to get the extra coverage.

File: src/ILLink.RoslynAnalyzer/TrimAnalysis/KnownValueType.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+using System.Diagnostics.CodeAnalysis;
+using System.Text;
+using ILLink.Shared;
+using ILLink.Shared.TrimAnalysis;
+using Microsoft.CodeAnalysis;
+
+namespace ILLink.RoslynAnalyzer.TrimAnalysis
+{
+	public record KnownValueType : ValueWithDynamicallyAccessedMembers
```
Comment (sbomer) [inline src/ILLink.RoslynAnalyzer/TrimAnalysis/KnownValueType.cs:12]: This can now use `SystemTypeValue`

File: src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs [hunk 2]
```diff
@@ -153,6 +178,56 @@ static DiagnosticId GetDiagnosticId (SymbolValue source, SymbolValue target)
...
+				member.GetDisplayName (),
+				message,
+				url);
+		}
+
+		static IEnumerable<Diagnostic> VerifyMethodSymbolForDiagnostic (IMethodSymbol methodSymbol, Location location)
+		{
+			foreach (var parameter in methodSymbol.Parameters) {
+				if (parameter.GetDynamicallyAccessedMemberTypes () != DynamicallyAccessedMemberTypes.None)
+					yield return Diagnostic.Create (DiagnosticDescriptors.GetDiagnosticDescriptor (DiagnosticId.DynamicallyAccessedMembersMethodAccessedViaReflection), location, methodSymbol.GetDisplayName ());
```
Comment (sbomer) [inline src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs:197]: I think we should avoid yielding multiple diagnostics - there should be just one diagnostic for the method if any of its parameters or the return type on a virtual have annotations.

Comment (tlakollo) [reply src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs:178]: I created a piece of code that I actually didnt like much which does yield return and then yield break to keep the IEnumerator and use common code. Maybe I should divide the Property that yields an IEnumerator from the method and field analysis which should return only one diagnostic. Let me know what you think

**Comment by tlakollo** [comment]
I will change this to draft, I realized that there is several tests having problems with interfaces and implementations. Need to spend more time on it.

File: test/Mono.Linker.Tests.Cases/RequiresCapability/RequiresAttributeMismatch.cs [hunk 3]
```diff
@@ -21,18 +21,30 @@ namespace Mono.Linker.Tests.Cases.RequiresCapability
...
-		[ExpectedWarning ("IL2026", "VirtualPropertyAnnotationInPropertyAndAccessor.get", ProducedBy = ProducedBy.Trimmer)]
-		[ExpectedWarning ("IL2026", "VirtualPropertyAnnotationInPropertyAndAccessor.get", ProducedBy = ProducedBy.Trimmer)]
-		[ExpectedWarning ("IL2026", "PropertyAnnotationInPropertyAndAccessor.set", ProducedBy = ProducedBy.Trimmer)]
+		[ExpectedWarning ("IL2026", "BaseClassWithRequires.VirtualPropertyAnnotationInAccesor.get")]
+		[ExpectedWarning ("IL2026", "BaseClassWithRequires.VirtualPropertyAnnotationInAccesor.get")]
+		[ExpectedWarning ("IL2026", "BaseClassWithRequires.VirtualPropertyAnnotationInAccesor.get")]
+		[ExpectedWarning ("IL2026", "DerivedClassWithRequires.VirtualPropertyAnnotationInAccesor.get", ProducedBy = ProducedBy.Analyzer)]
+		[ExpectedWarning ("IL2026", "DerivedClassWithAllWarnings.VirtualPropertyAnnotationInAccesor.set", ProducedBy = ProducedBy.Analyzer)]
+		[ExpectedWarning ("IL2026", "DerivedClassWithAllWarnings.VirtualPropertyAnnotationInProperty.get", ProducedBy = ProducedBy.Analyzer)]
+		[ExpectedWarning ("IL2026", "DerivedClassWithAllWarnings.VirtualPropertyAnnotationInProperty.set", ProducedBy = ProducedBy.Analyzer)]
```
Comment (vitek-karas) [inline test/Mono.Linker.Tests.Cases/RequiresCapability/RequiresAttributeMismatch.cs:30]: Do we have a bug for this discrepancy?

File: test/Mono.Linker.Tests.Cases/DataFlow/AnnotatedMembersAccessedViaReflection.cs [hunk 4]
```diff
@@ -643,11 +652,11 @@ struct ValueWithAnnotatedField
 				public Type _typeField;
 			}
 
-			[ExpectedWarning ("IL2110", nameof (ValueWithAnnotatedField._typeField))]
+			[ExpectedWarning ("IL2110", nameof (ValueWithAnnotatedField._typeField), ProducedBy = ProducedBy.Trimmer)]
 			[DllImport ("nonexistent")]
 			static extern ValueWithAnnotatedField GetValueWithAnnotatedField ();
 
-			[ExpectedWarning ("IL2110", nameof (ValueWithAnnotatedField._typeField))]
+			[ExpectedWarning ("IL2110", nameof (ValueWithAnnotatedField._typeField), ProducedBy = ProducedBy.Trimmer)]
```
Comment (vitek-karas) [inline test/Mono.Linker.Tests.Cases/DataFlow/AnnotatedMembersAccessedViaReflection.cs:659]: Ideally we should add a link to a bug - it will be easier to find.

File: test/Mono.Linker.Tests.Cases/DataFlow/AnnotatedMembersAccessedViaReflection.cs [hunk 5]
```diff
@@ -264,13 +264,22 @@ public static void Test ()
 		class AnnotatedMethodReturnValue
 		{
 			[return: DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.PublicMethods)]
-			public static Type StaticMethodWithAnnotatedReturnValue () => null;
+			public static Type StaticMethodWithAnnotatedReturnValue ()
+			{
+				return null;
+			}
```
Comment (vitek-karas) [inline test/Mono.Linker.Tests.Cases/DataFlow/AnnotatedMembersAccessedViaReflection.cs:270]: I don't really mind, but why did you change this?

File: test/ILLink.RoslynAnalyzer.Tests/DynamicallyAccessedMembersAnalyzerTests.cs [hunk 6]
```diff
@@ -160,17 +160,12 @@ private static void M(Type type)
...
 			// (17,9): warning IL2070: 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicMethods' in call to 'System.Type.GetMethod(String)'.
 			// The parameter 'type' of method 'C.M(Type)' does not have matching annotations.
 			// The source value must declare at least the same requirements as those declared on the target location it is assigned to.
-			return VerifyDynamicallyAccessedMembersAnalyzer (TargetMethodWithAnnotations,
-				VerifyCS.Diagnostic (DiagnosticId.DynamicallyAccessedMembersMismatchParameterTargetsThisParameter)
-				.WithSpan (17, 9, 17, 30)
-				.WithArguments ("System.Type.GetMethod(String)",
-					"type",
-					"C.M(Type)",
-					"'DynamicallyAccessedMemberTypes.PublicMethods'"));
```
Comment (vitek-karas) [inline test/ILLink.RoslynAnalyzer.Tests/DynamicallyAccessedMembersAnalyzerTests.cs:173]: Could you maybe just comment out the code - we should add it later once the intrinsics are implemented, right?

File: src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs [hunk 7]
```diff
@@ -38,6 +38,24 @@ public bool Invoke (MethodProxy calledMethod, MultiValue instanceValue, IReadOnl
...
+				return true;
+
+			case IntrinsicId.Type_GetConstructor:
+				return true;
+
+			case IntrinsicId.Type_GetEvent:
+				return true;
+
+			case IntrinsicId.Activator_CreateInstance_Type:
+				return true;
```
Comment (vitek-karas) [inline src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs:57]: Please add a comment why you're doing this.

File: test/Mono.Linker.Tests.Cases/RequiresCapability/RequiresAttributeMismatch.cs (hunk 3 above)
Comment (tlakollo) [reply test/Mono.Linker.Tests.Cases/RequiresCapability/RequiresAttributeMismatch.cs:30]: https://github.com/dotnet/linker/issues/2533
I will add links to issues to several places

File: src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionMethodBodyScanner.cs [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+			var message = RequiresUnreferencedCodeUtils.GetMessageFromAttribute (requiresAttributeData);
+			var url = RequiresAnalyzerBase.GetUrlFromAttribute (requiresAttributeData);
+			diagnosticContext.ReportDiagnostic (DiagnosticId.RequiresUnreferencedCode, member.GetDisplayName (), message, url);
+		}
+
+		static void MarkMethod (DiagnosticContext diagnosticContext, IMethodSymbol methodSymbol)
+		{
+			if (methodSymbol.TargetHasRequiresUnreferencedCodeAttribute (out var requiresAttributeData)
+						&& RequiresUnreferencedCodeUtils.VerifyRequiresUnreferencedCodeAttributeArguments (requiresAttributeData)) {
+				ReportRequiresUnreferencedCodeDiagnostic (diagnosticContext, requiresAttributeData, methodSymbol);
```
[the remark this answers is not in the document]
Comment (tlakollo) [reply src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionMethodBodyScanner.cs:55]: No, because property is not allowed to have RUC, only DAM is allowed

File: test/Mono.Linker.Tests.Cases/DataFlow/AnnotatedMembersAccessedViaReflection.cs (hunk 5 above)
Comment (tlakollo) [reply test/Mono.Linker.Tests.Cases/DataFlow/AnnotatedMembersAccessedViaReflection.cs:270]: At some point the analyzer was failing on arrow expressions, now its fixed so Im getting back the arrows

File: src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionMethodBodyScanner.cs [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+using System.Diagnostics.CodeAnalysis;
+using ILLink.RoslynAnalyzer.DataFlow;
+using ILLink.Shared;
+using ILLink.Shared.TrimAnalysis;
+using Microsoft.CodeAnalysis;
+
+namespace ILLink.RoslynAnalyzer.TrimAnalysis
+{
+	class ReflectionMethodBodyScanner
```
[the remark this answers is not in the document]
Comment (tlakollo) [reply src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionMethodBodyScanner.cs]: Changed to diagnosticSolver, let me know what you think

[the remark this answers is not in the document]
Comment (tlakollo) [reply src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionMethodBodyScanner.cs]: Offline discussion  - changed the name to ReflectionAccessAnalyzer

File: src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionMethodBodyScanner.cs [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+			if (methodSymbol.IsVirtual && methodSymbol.GetDynamicallyAccessedMemberTypesOnReturnType () != DynamicallyAccessedMemberTypes.None)
+				diagnosticContext.ReportDiagnostic (DiagnosticId.DynamicallyAccessedMembersMethodAccessedViaReflection, methodSymbol.GetDisplayName ());
+			else if (methodSymbol.IsVirtual && methodSymbol.MethodKind is MethodKind.PropertyGet &&
+				(methodSymbol.GetDynamicallyAccessedMemberTypes () != DynamicallyAccessedMemberTypes.None ||
+				methodSymbol.GetDynamicallyAccessedMemberTypesOnAssociatedSymbol () != DynamicallyAccessedMemberTypes.None)) {
+				diagnosticContext.ReportDiagnostic (DiagnosticId.DynamicallyAccessedMembersMethodAccessedViaReflection, methodSymbol.GetDisplayName ());
+			} else if (methodSymbol.MethodKind is MethodKind.PropertySet &&
+				  (methodSymbol.GetDynamicallyAccessedMemberTypesOnReturnType () != DynamicallyAccessedMemberTypes.None ||
+				  methodSymbol.GetDynamicallyAccessedMemberTypesOnAssociatedSymbol () != DynamicallyAccessedMemberTypes.None)) {
+				diagnosticContext.ReportDiagnostic (DiagnosticId.DynamicallyAccessedMembersMethodAccessedViaReflection, methodSymbol.GetDisplayName ());
```
[the remark this answers is not in the document]
Comment (tlakollo) [reply src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionMethodBodyScanner.cs]: I think I need to spend more time on this one, and probably add more tests.

File: src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionMethodBodyScanner.cs [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+					MarkProperty (diagnosticContext, property);
+					break;
+				/* Skip Type and InterfaceImplementation marking since doesnt seem relevant for diagnostic generation
+				case ITypeSymbol nestedType:
+					MarkType (diagnosticContext, nestedType);
+					break;
+				case InterfaceImplementation interfaceImplementation:
+					MarkInterfaceImplementation (analysisContext, interfaceImplementation, dependencyKind);
+					break;
+				*/
```
[the remark this answers is not in the document]
Comment (tlakollo) [reply src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionMethodBodyScanner.cs]: So I think is kind of handled, each member looks for RUC on the member and on the type based on the conditions (being static or constructor and the ContainingType has RUC), so at least type-level RUC seems fine to me. Maybe once more intrinsics and we enable more tests like [ObjectGetType](https://github.com/dotnet/linker/blob/main/test/Mono.Linker.Tests.Cases/Reflection/ObjectGetType.cs) we can have enable the DAM on type level

File: src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionAccessAnalyzer.cs [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+				case IMethodSymbol method:
+					GetDiagnosticsForMethod (diagnosticContext, method);
+					break;
+				case IFieldSymbol field:
+					GetDiagnosticsForField (diagnosticContext, field);
+					break;
+				case IPropertySymbol property:
+					GetDiagnosticsForProperty (diagnosticContext, property);
+					break;
+				/* Skip Type and InterfaceImplementation marking since doesnt seem relevant for diagnostic generation
```
[the remark this answers is not in the document]
Comment (vitek-karas) [reply src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionAccessAnalyzer.cs:28]: I think https://github.com/dotnet/linker/blob/3f704bb49eb2fb2847ceaecfbabe8d5f3e0c86c9/test/Mono.Linker.Tests.Cases/Reflection/TypeHierarchyReflectionWarnings.cs#L267 will cover this once we get there.

File: src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionAccessAnalyzer.cs [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+				diagnosticContext.ReportDiagnostic (DiagnosticId.DynamicallyAccessedMembersMethodAccessedViaReflection, methodSymbol.GetDisplayName ());
+			else if (methodSymbol.IsVirtual && methodSymbol.MethodKind is MethodKind.PropertyGet &&
+				(methodSymbol.GetDynamicallyAccessedMemberTypes () != DynamicallyAccessedMemberTypes.None ||
+				methodSymbol.GetDynamicallyAccessedMemberTypesOnAssociatedSymbol () != DynamicallyAccessedMemberTypes.None)) {
+				diagnosticContext.ReportDiagnostic (DiagnosticId.DynamicallyAccessedMembersMethodAccessedViaReflection, methodSymbol.GetDisplayName ());
+			} else if (methodSymbol.MethodKind is MethodKind.PropertySet &&
+				  (methodSymbol.GetDynamicallyAccessedMemberTypesOnReturnType () != DynamicallyAccessedMemberTypes.None ||
+				  methodSymbol.GetDynamicallyAccessedMemberTypesOnAssociatedSymbol () != DynamicallyAccessedMemberTypes.None)) {
+				diagnosticContext.ReportDiagnostic (DiagnosticId.DynamicallyAccessedMembersMethodAccessedViaReflection, methodSymbol.GetDisplayName ());
+			} else {
```
Comment (vitek-karas) [inline src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionAccessAnalyzer.cs:65]: I'm still not sure about this one.

```C#
class TestType
{
    [DAM(Methods)]
    public Type MyType { get; }

    public Type MyTypeWithGetter { [return: DAM(Methods)] get { return null; } }
}

// This should produce 2 warnings
// but the code above will very likely only produce 1 - the first one, not for `MyTypeWithGetter`
typeof (TestType).RequiresAll ();
```

The reason is that on a getter we check if the method itself has DAM, but on a setter we check if the return value has DAM. Return value on setter is void, so that is super weird as well.

**Comment by tlakollo** [comment]
Added a bug for most of the properties inconsistencies https://github.com/dotnet/linker/issues/2575
The problem is naming so I'm not sure if it's linker logic or test infra. The system to present warnings in those cases is wrong in many ways. For example, makes a non virtual getter to warn

File: src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionAccessAnalyzer.cs (hunk 13 above)
Comment (tlakollo) [reply src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionAccessAnalyzer.cs]: Added test for these cases

File: src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionMethodBodyScanner.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (tlakollo) [reply src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionMethodBodyScanner.cs]: End up using the FlowAnnotations helpers by Vitek suggestion. Thanks Vitek for the review :D

File: test/ILLink.RoslynAnalyzer.Tests/DynamicallyAccessedMembersAnalyzerTests.cs [hunk 14]
```diff
@@ -160,17 +160,19 @@ private static void M(Type type)
...
+			return VerifyDynamicallyAccessedMembersAnalyzer (TargetMethodWithAnnotations
+							/*,
+							VerifyCS.Diagnostic (DiagnosticId.DynamicallyAccessedMembersMismatchParameterTargetsThisParameter)
+							.WithSpan (17, 9, 17, 30)
+							.WithArguments ("System.Type.GetMethod(String)",
+								"type",
+								"C.M(Type)",
+								"'DynamicallyAccessedMemberTypes.PublicMethods'")*/);
```
Comment (tlakollo) [inline test/ILLink.RoslynAnalyzer.Tests/DynamicallyAccessedMembersAnalyzerTests.cs:175]: Not sure why lint decided on this format but is on purpose

File: src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionAccessAnalyzer.cs [hunk 15]
```diff
@@ -44,29 +44,23 @@ static void ReportRequiresUnreferencedCodeDiagnostic (in DiagnosticContext diagn
...
-				(methodSymbol.GetDynamicallyAccessedMemberTypes () != DynamicallyAccessedMemberTypes.None ||
-				methodSymbol.GetDynamicallyAccessedMemberTypesOnAssociatedSymbol () != DynamicallyAccessedMemberTypes.None)) {
-				diagnosticContext.ReportDiagnostic (DiagnosticId.DynamicallyAccessedMembersMethodAccessedViaReflection, methodSymbol.GetDisplayName ());
-			} else if (methodSymbol.MethodKind is MethodKind.PropertySet &&
-				  (methodSymbol.GetDynamicallyAccessedMemberTypesOnReturnType () != DynamicallyAccessedMemberTypes.None ||
-				  methodSymbol.GetDynamicallyAccessedMemberTypesOnAssociatedSymbol () != DynamicallyAccessedMemberTypes.None)) {
-				diagnosticContext.ReportDiagnostic (DiagnosticId.DynamicallyAccessedMembersMethodAccessedViaReflection, methodSymbol.GetDisplayName ());
-			} else {
+			if (methodSymbol.GetDynamicallyAccessedMemberTypes () != DynamicallyAccessedMemberTypes.None)
+				diagnosticContext.AddDiagnostic (DiagnosticId.DynamicallyAccessedMembersMethodAccessedViaReflection, methodSymbol.GetDisplayName ());
```
Comment (vitek-karas) [inline src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionAccessAnalyzer.cs:56]: Ideally this should be only done on instance method (non-static) and only when the declaring type derives from `System.Type`. The latter part is tracked by either https://github.com/dotnet/linker/issues/2409 or a line item in https://github.com/dotnet/linker/issues/2273.

So probably OK to leave it as is for now - as we will likely handle this as a separate part of the analyzer. It might produce noise if there's such a case in the code though, but not a big deal.

File: src/ILLink.Shared/TrimAnalysis/RequireDynamicallyAccessedMembersAction.cs [hunk 16]
```diff
@@ -22,7 +22,7 @@ public void Invoke (in DiagnosticContext diagnosticContext, in MultiValue value,
 					var availableMemberTypes = valueWithDynamicallyAccessedMembers.DynamicallyAccessedMemberTypes;
 					if (!Annotations.SourceHasRequiredAnnotations (availableMemberTypes, targetValue.DynamicallyAccessedMemberTypes, out var missingMemberTypes)) {
 						(var diagnosticId, var diagnosticArguments) = Annotations.GetDiagnosticForAnnotationMismatch (valueWithDynamicallyAccessedMembers, targetValue, missingMemberTypes);
-						diagnosticContext.AddDiagnostic (diagnosticId, diagnosticArguments);
+						_diagnosticContext.AddDiagnostic (diagnosticId, diagnosticArguments);
```
Comment (vitek-karas) [inline src/ILLink.Shared/TrimAnalysis/RequireDynamicallyAccessedMembersAction.cs:25]: Declare the `_diagnosticContext` field on the shared struct (so in this file) and as readonly. That will "force" the analyzer/linker to initialize it in the .ctors. Cleaner than relying on the fields existence from the "outside".

File: test/Mono.Linker.Tests.Cases/DataFlow/AnnotatedMembersAccessedViaReflection.cs [hunk 17]
```diff
@@ -470,10 +521,26 @@ static void DynamicallyAccessedMembersSuppressedByRUC ()
...
+			[ExpectedWarning ("IL2111", nameof (PropertyWithAnnotationOnMembers) + ".set")]
+			[ExpectedWarning ("IL2111", nameof (PropertyWithAnnotationOnMembers) + ".set")]
+			[ExpectedWarning ("IL2111", nameof (PropertyWithAnnotationOnMembers) + ".set", ProducedBy = ProducedBy.Trimmer)]
+			[ExpectedWarning ("IL2111", nameof (PropertyWithAnnotationOnMembers) + ".set", ProducedBy = ProducedBy.Trimmer)]
+			[ExpectedWarning ("IL2111", nameof (PropertyWithAnnotationOnMembers) + ".get", ProducedBy = ProducedBy.Trimmer)]
+			[ExpectedWarning ("IL2111", nameof (PropertyWithAnnotationOnMembers) + ".get", ProducedBy = ProducedBy.Trimmer)]
+			[ExpectedWarning ("IL2111", nameof (VirtualPropertyWithAnnotationOnMembers) + ".get", ProducedBy = ProducedBy.Analyzer)]
+			[ExpectedWarning ("IL2111", nameof (VirtualPropertyWithAnnotationOnMembers) + ".get", ProducedBy = ProducedBy.Analyzer)]
+			[ExpectedWarning ("IL2111", nameof (VirtualPropertyWithAnnotationOnMembers) + ".set", ProducedBy = ProducedBy.Analyzer)]
+			[ExpectedWarning ("IL2111", nameof (VirtualPropertyWithAnnotationOnMembers) + ".set", ProducedBy = ProducedBy.Analyzer)]
```
[the remark this answers is not in the document]
Comment (vitek-karas) [reply test/Mono.Linker.Tests.Cases/DataFlow/AnnotatedMembersAccessedViaReflection.cs]: I pushed a change which fixes it for the property test - maybe we should do the same for the other tests in this file as well.

The fix is more of a workaround, I simply renamed the properties so that the substrings don't match.
As far as I can tell the correct warnings are generated. And the linker and analyzer behavior matches exactly now (nice).

That said there are still duplicate warnings, but that's a different matter.

File: src/ILLink.RoslynAnalyzer/TrimAnalysis/RequireDynamicallyAccessedMembersAction.cs [hunk 18]
```diff
@@ -7,9 +7,9 @@
 
 namespace ILLink.Shared.TrimAnalysis
 {
+#pragma warning disable CS0282 // There is no defined ordering between fields in multiple declarations of partial struct
```
Comment (vitek-karas) [inline src/ILLink.RoslynAnalyzer/TrimAnalysis/RequireDynamicallyAccessedMembersAction.cs:10]: Cleaner way is to add `[StructLayout (LayoutKind.Auto)]` on the struct - it will shut down the warning and also tells the runtime it's free to optimize the layout if it helps.

Comment (vitek-karas) [reply src/ILLink.RoslynAnalyzer/TrimAnalysis/RequireDynamicallyAccessedMembersAction.cs:10]: Also please add onto the shared source.

File: src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionAccessAnalyzer.cs [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+			if (eventSymbol.RemoveMethod is not null)
+				GetDiagnosticsForMethod (diagnosticContext, eventSymbol.RemoveMethod);
+			if (eventSymbol.RaiseMethod is not null)
+				GetDiagnosticsForMethod (diagnosticContext, eventSymbol.RaiseMethod);
+		}
+
+		static void GetDiagnosticsForField (in DiagnosticContext diagnosticContext, IFieldSymbol fieldSymbol)
+		{
+			if (fieldSymbol.TryGetRequiresUnreferencedCodeAttribute (out var requiresAttributeData))
+				ReportRequiresUnreferencedCodeDiagnostic (diagnosticContext, requiresAttributeData, fieldSymbol);
```
Comment (vitek-karas) [inline src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionAccessAnalyzer.cs:90]: I don't think we allows RUC on a field, right?

**Review by vitek-karas — APPROVED** [review]
Looks good - just a few minor things

File: src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionAccessAnalyzer.cs (hunk 19 above)
Comment (tlakollo) [reply src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionAccessAnalyzer.cs]: No, we don't - deleted the code

File: src/ILLink.RoslynAnalyzer/TrimAnalysis/RequireDynamicallyAccessedMembersAction.cs [hunk 20]
```diff
@@ -2,12 +2,13 @@
...
 
 using System.Diagnostics.CodeAnalysis;
+using System.Runtime.InteropServices;
 using ILLink.RoslynAnalyzer.TrimAnalysis;
 using ILLink.Shared.TypeSystemProxy;
 
 namespace ILLink.Shared.TrimAnalysis
 {
-#pragma warning disable CS0282 // There is no defined ordering between fields in multiple declarations of partial struct
+	[StructLayout (LayoutKind.Auto)]
```
Comment (vitek-karas) [inline src/ILLink.RoslynAnalyzer/TrimAnalysis/RequireDynamicallyAccessedMembersAction.cs:11]: Nit: put this onto the shared source file - it will apply to both analyzer and linker.

File: src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionAccessAnalyzer.cs (hunk 19 above)
Comment (tlakollo) [reply src/ILLink.RoslynAnalyzer/TrimAnalysis/ReflectionAccessAnalyzer.cs]: Apparently, we are when RequiresUnreferencedCode is on a class, I think we mark access to fields as dangerous because they can call the static constructor. Getting back the code

**Review by vitek-karas — APPROVED** [review]
:shipit:
